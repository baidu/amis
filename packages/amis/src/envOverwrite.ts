/**
 * @file 用于在移动端或不同语言环境下使用不同配置
 */

import {SchemaNode, Schema} from './types';
import {RendererProps, RendererConfig, addSchemaFilter} from './factory';
import {findObjectsWithKey} from './utils/helper';
const isMobile = (window as any).matchMedia?.('(max-width: 768px)').matches
  ? true
  : false;

// 这里不能用 addSchemaFilter 是因为还需要更深层的替换，比如 select 里的 options
export const envOverwrite = (schema: any, locale?: string) => {
  if (schema.mobile && isMobile) {
    Object.assign(schema, schema.mobile);
    delete schema.mobile;
  }

  if (locale) {
    let schemaNodes = findObjectsWithKey(schema, locale);
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode[locale]);
      delete schemaNode[locale];
    }
  }

  if (isMobile) {
    let schemaNodes = findObjectsWithKey(schema, 'mobile');
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode['mobile']);
      delete schemaNode['mobile'];
    }
  }
};
