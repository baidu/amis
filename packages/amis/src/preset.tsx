import {
  addRootWrapper,
  extendDefaultEnv,
  LazyComponent,
  render,
  themeable,
  ThemeProps
} from 'amis-core';

import {alert, confirm, setRenderSchemaFn, toast, ImageGallery} from 'amis-ui';
import React from 'react';

extendDefaultEnv({
  alert,
  confirm,
  notify: (type: keyof typeof toast, msg: string, conf: any) =>
    toast[type] ? toast[type](msg, conf) : console.warn('[Notify]', type, msg)
});

setRenderSchemaFn((controls, value, callback, scopeRef, theme) => {
  return render(
    {
      name: 'form',
      type: 'form',
      wrapWithPanel: false,
      mode: 'horizontal',
      controls,
      messages: {
        validateFailed: ''
      }
    },
    {
      data: value,
      onFinished: callback,
      scopeRef,
      theme
    },
    {
      session: 'prompt'
    }
  );
});

addRootWrapper((props: any) => {
  const {env, children} = props;
  return (
    <ImageGallery modalContainer={env.getModalContainer}>
      {children}
    </ImageGallery>
  );
});

const SimpleSpinner = themeable(
  (
    props: {
      className?: string;
      spinnerClassName?: string;
    } & ThemeProps
  ) => {
    const cx = props.classnames;

    return (
      <div
        data-testid="spinner"
        className={cx(`Spinner`, 'in', props.className)}
      >
        <div
          className={cx(
            `Spinner-icon`,
            'Spinner-icon--default',
            props.spinnerClassName
          )}
        ></div>
      </div>
    );
  }
);

LazyComponent.defaultProps.placeholder = <SimpleSpinner />;
