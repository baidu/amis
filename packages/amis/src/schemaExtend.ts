/**
 * 扩展 Schema，目前用于实现 input-kv
 */
import isEqual from 'lodash/isEqual';
import {Schema} from 'amis-core';
import {addSchemaFilter} from 'amis-core';
import {isObject} from 'amis-core';

// input-kv 实际上是 combo 的一种扩展
addSchemaFilter(function (schema: Schema, renderer, props?: any) {
  if (schema && schema.type === 'input-kv') {
    const autoParseJSON = schema.autoParseJSON ?? true;
    return {
      draggable: true,
      ...schema,
      multiple: true,
      pipeIn: function (this: any, value: any) {
        if (!isObject(value)) {
          return [];
        }
        if (isEqual(value, this.cachedValue)) {
          return this.cachedValueArray;
        }
        const arr: Array<any> = [];
        Object.keys(value).forEach(key => {
          const valueType = typeof value[key];
          arr.push({
            key: key || '',
            value:
              valueType === 'string' ||
              valueType === 'number' ||
              valueType === 'boolean'
                ? value[key]
                : JSON.stringify(value[key])
          });
        });
        return arr;
      },
      pipeOut: function (this: any, value: any) {
        if (!Array.isArray(value)) {
          return value;
        }
        const obj: any = {};
        value.forEach((item: any) => {
          const key: string = item.key ?? '';
          let value: any = item.value ?? schema.defaultValue ?? '';
          if (
            autoParseJSON &&
            typeof value === 'string' &&
            value.startsWith('{')
          ) {
            try {
              value = JSON.parse(value);
            } catch (e) {}
          }

          obj[key] = value;
        });
        this.cachedValue = obj;
        this.cachedValueArray = value;
        return obj;
      },
      items: [
        schema.keySchema
          ? {
              ...schema.keySchema,
              unique: true,
              name: 'key',
              required: true,
              validateOnChange: true
            }
          : {
              placeholder: schema.keyPlaceholder ?? 'Key',
              type: 'input-text',
              unique: true,
              name: 'key',
              required: true,
              validateOnChange: true
            },
        schema.valueSchema
          ? {...schema.valueSchema, name: 'value'}
          : schema.valueComponent
          ? {
              placeholder: schema.valuePlaceholder ?? 'Value',
              component: schema.valueComponent,
              asFormItem: true,
              name: 'value'
            }
          : {
              placeholder: schema.valuePlaceholder ?? 'Value',
              type: schema.valueType || 'input-text',
              name: 'value'
            }
      ]
    };
  }

  return schema;
});

/**
 * 之前 input-kv 的 value 值不支持对象或数组
 * 很多属性是给单个值设置的，比如 valuePlaceholder 导致
 * 耦合在一起会导致配置项混乱，所以新增了这个组件专门支持 value 是对象或数组的场景
 */
addSchemaFilter(function (schema: Schema, renderer, props?: any) {
  if (schema && schema.type === 'input-kvs') {
    const keyItem = schema.keyItem || {};
    const valueItems = schema.valueItems || [];
    // value 直接放在 key 下的情况
    let flatValue = false;
    if (valueItems.length == 1) {
      if (valueItems[0].name === '_value') {
        flatValue = true;
      }
    }
    const newSchema = {
      draggable: true,
      multiple: true,
      multiLine: true,
      ...schema,
      pipeIn: (data: any) => {
        if (!isObject(data)) {
          return [];
        }
        const arr: Array<any> = [];
        Object.keys(data).forEach(key => {
          let value = data[key];
          if (!value) {
            // value = {};
          }
          if (flatValue) {
            arr.push({
              _key: key || '',
              _value: value
            });
          } else if (typeof value === 'object') {
            arr.push({
              ...value,
              _key: key || ''
            });
          }
        });
        return arr;
      },
      pipeOut: (value: any) => {
        if (!Array.isArray(value)) {
          return value;
        }
        const obj: any = {};
        value.forEach((item: any) => {
          let {_key, ...rest} = item;
          _key = _key ?? '';
          if (flatValue) {
            if (schema.valueIsArray) {
              obj[_key] = item['_value'] || [];
            } else {
              obj[_key] = item['_value'] || {};
            }
            // 数组的时候初始化会生成 [{}]，还不确定是哪生成的，先修正为 []
            if (isEqual(obj[_key], [{}])) {
              obj[_key] = [];
            }
          } else {
            if (schema.valueIsArray) {
              obj[_key] = rest || [];
            } else {
              obj[_key] = rest || {};
            }
          }
        });
        return obj;
      },
      items: [
        {
          type: 'input-text',
          unique: true,
          name: '_key',
          required: true,
          validateOnChange: true,
          ...keyItem
        },
        ...valueItems
      ]
    };
    return newSchema;
  }

  return schema;
});
