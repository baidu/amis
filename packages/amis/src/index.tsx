/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */
export * from 'amis-core';
export * from 'amis-ui';

import './preset';

// 注册渲染器
import './renderers/Action';
import './renderers/Alert';
import './renderers/App';
import './renderers/Avatar';
import './renderers/Remark';
import './renderers/ButtonGroup';
import './renderers/Form/ButtonToolbar';
import './renderers/Breadcrumb';
import './renderers/DropDownButton';
import './renderers/Calendar';
import './renderers/Collapse';
import './renderers/CollapseGroup';
import './renderers/Color';
import './renderers/CRUD';
import './renderers/CRUD2';
import './renderers/Pagination';
import './renderers/Cards';
import './renderers/Card';
import './renderers/Card2';
import './renderers/Custom';
import './renderers/Date';
import './renderers/Dialog';
import './renderers/Divider';
import './renderers/Each';
import './renderers/Flex';
import './renderers/Form/Control';
import './renderers/Form/Hidden';
import './renderers/Form/InputText';
import './renderers/Form/InputTag';
import './renderers/Form/InputNumber';
import './renderers/Form/Textarea';
import './renderers/Form/Checkboxes';
import './renderers/Form/Checkbox';
import './renderers/Form/InputCity';
import './renderers/Form/ChartRadios';
import './renderers/Form/InputRating';
import './renderers/Form/Switch';
import './renderers/Form/Radios';
import './renderers/Form/Radio';
import './renderers/Form/JSONSchema';
import './renderers/Form/JSONSchemaEditor';
import './renderers/Form/ListSelect';
import './renderers/Form/LocationPicker';
import './renderers/Form/Select';
import './renderers/Form/Static';
import './renderers/Form/InputDate';
import './renderers/Form/InputDateRange';
import './renderers/Form/InputFormula';
import './renderers/Form/InputRepeat';
import './renderers/Form/InputTree';
import './renderers/Form/TreeSelect';
import './renderers/Form/InputImage';
import './renderers/Form/InputFile';
import './renderers/Form/UUID';
import './renderers/Form/MatrixCheckboxes';
import './renderers/Form/InputMonthRange';
import './renderers/Form/InputQuarterRange';
import './renderers/Form/InputYearRange';
import './renderers/Form/InputRange';
import './renderers/Form/InputArray';
import './renderers/Form/Combo';
import './renderers/Form/ConditionBuilder';
import './renderers/Form/InputSubForm';
import './renderers/Form/InputExcel';
import './renderers/Form/InputRichText';
import './renderers/Form/Editor';
import './renderers/Form/DiffEditor';
import './renderers/Form/InputColor';
import './renderers/Form/ChainedSelect';
import './renderers/Form/NestedSelect';
import './renderers/Form/Transfer';
import './renderers/Form/TransferPicker';
import './renderers/Form/InputTable';
import './renderers/Form/Picker';
import './renderers/Form/IconPicker';
import './renderers/Form/IconSelect';
import './renderers/Form/Formula';
import './renderers/Form/FieldSet';
import './renderers/Form/TabsTransfer';
import './renderers/Form/TabsTransferPicker';
import './renderers/Form/Group';
import './renderers/Form/InputGroup';
import './renderers/Form/UserSelect';
import './renderers/Grid';
import './renderers/Grid2D';
import './renderers/HBox';
import './renderers/VBox';
import './renderers/Image';
import './renderers/Images';
import './renderers/List';
import './renderers/Log';
import './renderers/Operation';
import './renderers/Page';
import './renderers/PaginationWrapper';
import './renderers/Panel';
import './renderers/Plain';
import './renderers/Property';
import './renderers/Portlet';
import './renderers/Spinner';
import './renderers/Table/index';
import './renderers/Tabs';
import './renderers/Tpl';
import './renderers/Mapping';
import './renderers/Progress';
import './renderers/Status';
import './renderers/Json';
import './renderers/Link';
import './renderers/Wizard';
import './renderers/Chart';
import './renderers/Container';
import './renderers/SwitchContainer';
import './renderers/SearchBox';
import './renderers/Service';
import './renderers/SparkLine';
import './renderers/Video';
import './renderers/Audio';
import './renderers/Nav';
import './renderers/Number';
import './renderers/Tasks';
import './renderers/Drawer';
import './renderers/Wrapper';
import './renderers/IFrame';
import './renderers/BarCode';
import './renderers/QRCode';
import './renderers/Icon';
import './renderers/Carousel';
import './renderers/AnchorNav';
import './renderers/Steps';
import './renderers/Timeline';
import './renderers/Markdown';
import './renderers/TableView';
import './renderers/Code';
import './renderers/WebComponent';
import './renderers/GridNav';
import './renderers/TooltipWrapper';
import './renderers/Tag';
import './renderers/Table2/index';
import './renderers/Words';
import './renderers/Password';
import './renderers/DateRange';
import './renderers/MultilineText';
import './renderers/OfficeViewer';
import './renderers/AMIS';

import './compat';
import './schemaExtend';
import type {
  BaseSchema,
  FormSchema,
  SchemaApi,
  SchemaCollection,
  SchemaExpression,
  SchemaObject
} from './Schema';
import type {TableViewSchema, TrObject} from './renderers/TableView';
import type {ActionSchema, ButtonSchema} from './renderers/Action';
import type {CRUDCommonSchema} from './renderers/CRUD';
import type {CRUD2Schema} from './renderers/CRUD2';
import type {TabsSchema} from './renderers/Tabs';
import {availableLanguages as EditorAvailableLanguages} from './renderers/Form/Editor';
import type {Action} from './types';
export * from './renderers/Form/IconPickerIcons';
export * from './renderers/Form/IconSelectStore';

export {
  BaseSchema,
  SchemaCollection,
  FormSchema,
  SchemaApi,
  SchemaObject,
  TableViewSchema,
  TrObject,
  ActionSchema,
  CRUDCommonSchema,
  ButtonSchema,
  CRUD2Schema,
  TabsSchema,
  SchemaExpression,
  Action,
  EditorAvailableLanguages
};
