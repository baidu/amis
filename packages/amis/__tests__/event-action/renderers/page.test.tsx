import React = require('react');
import * as renderer from 'react-test-renderer';
import {fireEvent, render, waitFor} from '@testing-library/react';
import '../../../src';
import {render as amisRender} from '../../../src';
import {makeEnv} from '../../helper';

test('doAction:page reload', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 0,
        msg: 'ok',
        data: {
          title: 'Test Page Component',
          date: '2017-10-13'
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'page',
        initApi: '/api/mock2/page/initData',
        id: 'page_reload',
        body: [
          {
            type: 'button',
            label: '刷新Page数据加载请求',
            level: 'primary',
            className: 'mb-2',
            onEvent: {
              click: {
                actions: [
                  {
                    componentId: 'page_reload',
                    actionType: 'reload'
                  }
                ]
              }
            }
          },
          {
            type: 'input-text',
            label: '我的标题：',
            name: 'title'
          },
          {
            type: 'input-text',
            label: '我的年龄：',
            name: 'age'
          }
        ]
      },
      {},
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  await waitFor(() => {
    expect(getByText('刷新Page数据加载请求')).toBeInTheDocument();
    expect(
      container.querySelector('[name="title"][value="Test Page Component"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="age"][value=""]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[data-testid="spinner"]')
    ).not.toBeInTheDocument();
  });

  expect(container).toMatchSnapshot();

  fireEvent.click(getByText(/刷新Page数据加载请求/));

  await waitFor(() => {
    expect(fetcher).toHaveBeenCalledTimes(2);
    expect(fetcher.mock.calls[0][0].url).toEqual('/api/mock2/page/initData');
    expect(fetcher.mock.calls[1][0].url).toEqual('/api/mock2/page/initData');
    expect(
      container.querySelector('[name="title"][value="Test Page Component"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="age"][value=""]')
    ).toBeInTheDocument();
  });

  expect(container).toMatchSnapshot();
});

test('doAction:page reload with data', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 0,
        msg: 'ok',
        data: {
          title: 'Test Page Component',
          date: '2017-10-13'
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'page',
        initApi: '/api/mock2/page/initData',
        id: 'page_reload',
        body: [
          {
            type: 'button',
            label: '刷新Page数据加载请求，同时把年龄更新为18',
            level: 'primary',
            className: 'mb-2',
            onEvent: {
              click: {
                actions: [
                  {
                    componentId: 'page_reload',
                    actionType: 'reload',
                    data: {
                      age: '18'
                    }
                  }
                ]
              }
            }
          },
          {
            type: 'input-text',
            label: '我的标题：',
            name: 'title'
          },
          {
            type: 'input-text',
            label: '我的年龄：',
            name: 'age'
          }
        ]
      },
      {},
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  await waitFor(() => {
    expect(
      getByText('刷新Page数据加载请求，同时把年龄更新为18')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="title"][value="Test Page Component"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="age"][value=""]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[data-testid="spinner"]')
    ).not.toBeInTheDocument();
  });

  expect(container).toMatchSnapshot();

  fireEvent.click(getByText(/刷新Page数据加载请求，同时把年龄更新为18/));

  await waitFor(() => {
    expect(fetcher).toHaveBeenCalledTimes(2);
    expect(fetcher.mock.calls[0][0].url).toEqual('/api/mock2/page/initData');
    expect(fetcher.mock.calls[1][0].url).toEqual('/api/mock2/page/initData');
    expect(
      container.querySelector('[name="title"][value="Test Page Component"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="age"][value="18"]')
    ).toBeInTheDocument();
  });

  expect(container).toMatchSnapshot();
});

test('doAction:page reload with data(same name)', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 0,
        msg: 'ok',
        data: {
          title: 'Test Page Component',
          date: '2017-10-13'
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'page',
        initApi: '/api/mock2/page/initData',
        id: 'page_reload',
        body: [
          {
            type: 'button',
            label: '刷新Page数据加载请求，请求返回的date会覆盖动作传递的date',
            level: 'primary',
            className: 'mb-2',
            onEvent: {
              click: {
                actions: [
                  {
                    componentId: 'page_reload',
                    actionType: 'reload',
                    data: {
                      date: '2023-07-07'
                    }
                  }
                ]
              }
            }
          },
          {
            type: 'input-text',
            label: '我的标题：',
            name: 'title'
          },
          {
            type: 'input-text',
            label: '日期：',
            name: 'date'
          }
        ]
      },
      {},
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  await waitFor(() => {
    expect(
      getByText('刷新Page数据加载请求，请求返回的date会覆盖动作传递的date')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="title"][value="Test Page Component"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="date"][value="2017-10-13"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[data-testid="spinner"]')
    ).not.toBeInTheDocument();
  });

  expect(container).toMatchSnapshot();

  fireEvent.click(
    getByText(/刷新Page数据加载请求，请求返回的date会覆盖动作传递的date/)
  );

  await waitFor(() => {
    expect(fetcher).toHaveBeenCalledTimes(2);
    expect(fetcher.mock.calls[0][0].url).toEqual('/api/mock2/page/initData');
    expect(fetcher.mock.calls[1][0].url).toEqual('/api/mock2/page/initData');
    expect(
      container.querySelector('[name="title"][value="Test Page Component"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('[name="date"][value="2017-10-13"]')
    ).toBeInTheDocument();
  });

  expect(container).toMatchSnapshot();
});
