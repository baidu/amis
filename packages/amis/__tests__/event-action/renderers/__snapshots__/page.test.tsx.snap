// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doAction:page reload 1`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default mb-2"
            type="button"
          >
            <span>
              刷新Page数据加载请求
            </span>
          </button>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的标题：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="title"
                  placeholder=""
                  size="10"
                  type="text"
                  value="Test Page Component"
                />
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的年龄：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="age"
                  placeholder=""
                  size="10"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`doAction:page reload 2`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <div
            class="cxd-Spinner-overlay in"
          />
          <div
            class="cxd-Spinner cxd-Spinner--overlay in"
            data-testid="spinner"
          >
            <div
              class="cxd-Spinner-icon cxd-Spinner-icon--lg cxd-Spinner-icon--default"
            />
          </div>
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default mb-2"
            type="button"
          >
            <span>
              刷新Page数据加载请求
            </span>
          </button>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的标题：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="title"
                  placeholder=""
                  size="10"
                  type="text"
                  value="Test Page Component"
                />
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的年龄：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="age"
                  placeholder=""
                  size="10"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`doAction:page reload with data 1`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default mb-2"
            type="button"
          >
            <span>
              刷新Page数据加载请求，同时把年龄更新为18
            </span>
          </button>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的标题：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="title"
                  placeholder=""
                  size="10"
                  type="text"
                  value="Test Page Component"
                />
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的年龄：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="age"
                  placeholder=""
                  size="10"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`doAction:page reload with data 2`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <div
            class="cxd-Spinner-overlay in"
          />
          <div
            class="cxd-Spinner cxd-Spinner--overlay in"
            data-testid="spinner"
          >
            <div
              class="cxd-Spinner-icon cxd-Spinner-icon--lg cxd-Spinner-icon--default"
            />
          </div>
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default mb-2"
            type="button"
          >
            <span>
              刷新Page数据加载请求，同时把年龄更新为18
            </span>
          </button>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的标题：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="title"
                  placeholder=""
                  size="10"
                  type="text"
                  value="Test Page Component"
                />
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的年龄：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="age"
                  placeholder=""
                  size="10"
                  type="text"
                  value="18"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`doAction:page reload with data(same name) 1`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default mb-2"
            type="button"
          >
            <span>
              刷新Page数据加载请求，请求返回的date会覆盖动作传递的date
            </span>
          </button>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的标题：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="title"
                  placeholder=""
                  size="10"
                  type="text"
                  value="Test Page Component"
                />
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    日期：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="date"
                  placeholder=""
                  size="10"
                  type="text"
                  value="2017-10-13"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`doAction:page reload with data(same name) 2`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <div
            class="cxd-Spinner-overlay"
          />
          <div
            class="cxd-Spinner cxd-Spinner--overlay"
            data-testid="spinner"
          >
            <div
              class="cxd-Spinner-icon cxd-Spinner-icon--lg cxd-Spinner-icon--default"
            />
          </div>
          <button
            class="cxd-Button cxd-Button--primary cxd-Button--size-default mb-2"
            type="button"
          >
            <span>
              刷新Page数据加载请求，请求返回的date会覆盖动作传递的date
            </span>
          </button>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    我的标题：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="title"
                  placeholder=""
                  size="10"
                  type="text"
                  value="Test Page Component"
                />
              </div>
            </div>
          </div>
          <div
            class="cxd-Form-item cxd-Form-item--normal"
            data-role="form-item"
          >
            <label
              class="cxd-Form-label"
            >
              <span>
                <span
                  class="cxd-TplField"
                >
                  <span>
                    日期：
                  </span>
                </span>
              </span>
            </label>
            <div
              class="cxd-Form-control cxd-TextControl"
            >
              <div
                class="cxd-TextControl-input"
              >
                <input
                  autocomplete="off"
                  class=""
                  name="date"
                  placeholder=""
                  size="10"
                  type="text"
                  value="2017-10-13"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
